File: matlab/validateCostFunction.m
function validationResult = validateCostFunction(payloadJson)
    payload = jsondecode(payloadJson);

    func = payload.func;
    allowedSymbols = payload.allowedSymbols;
    interventions = payload.interventions;

    validationResult = struct();

    validationResult.isValid = true;
    validationResult.message = '';

    validateExpressionData = struct();
    validateExpressionData.expression = func;
    validateExpressionData.allowedSymbols = cellArray(allowedSymbols);

    validateExpressionResult = validateExpression(jsonencode(validateExpressionData));

    if ~validateExpressionResult.isValid
        validationResult = validateExpressionResult;

        return;
    end

    func = str2sym(func);

    % Get derivative of function and check whether all interventions are in the derivative
    for i = 1:length(interventions)
        intervention = interventions(i);

        derivative = diff(func, intervention);

        if ~contains(char(derivative), intervention)
            validationResult = struct('isValid', false, 'message', sprintf('Derivative of cost function does not contain intervention %s', char(intervention)));

            return;
        end
    end
end

File: backend/validateCostFunction.py
from sympy import diff, sympify, Symbol
from sympy.abc import _clash

from validateExpression import validateExpression


def validateCostFunction(payload):
    """Cost function validation function"""

    # Unpacking the payload
    func = payload['func']
    allowedSymbols = payload['allowedSymbols']
    interventions = payload['interventions']

    # Initializing the validation result
    validationResult = {
        'isValid': True,
        'message': ''
    }

    # Validating the expression
    validateExpressionData = {
        'expression': func,
        'allowedSymbols': allowedSymbols
    }

    # Validating the expression
    validateExpressionResult = validateExpression(validateExpressionData)

    # Checking if the expression is valid
    if not validateExpressionResult['isValid']:
        return validateExpressionResult

    # Validating the cost function
    symbolicFunction = sympify(func, evaluate=False, locals=_clash)

    # Checking if the cost function contains all interventions
    for intervention in interventions:
        interventionSymbol = Symbol(intervention)
        derivative = diff(symbolicFunction, interventionSymbol)

        if interventionSymbol not in derivative.free_symbols:
            validationResult['isValid'] = False
            validationResult[
                'message'] = f'Derivative of cost function does not contain intervention "{intervention}"'

            break

    return validationResult
